### PR TITLE
XTheme: Show edit link in model choice fields (SHOOP-2170)

### PR DESCRIPTION
**Pull request by tulimaki**
Add custom XThemeModelChoiceField and XThemeModelChoiceWidget for
showing edit link to admin

Refs SHOOP-1892 / SHOOP-2170

### CODE REVIEW

File: shoop/xtheme/plugins/widgets.py [hunk 1]
```diff
@@ -90,3 +91,38 @@ def decompress(self, value):
...
             out_dict[FALLBACK_LANGUAGE_CODE] = value
         return out_dict
+
+
+class XThemeModelChoiceWidget(forms.Select):
+    def render(self, name, value, attrs=None, choices=()):
+        return mark_safe(
+            render_to_string("shoop/xtheme/_model_widget.jinja", {
+                "name": name,
+                "selected_object": [object for pk, object in self.choices if pk == int(value)][0] if value else None,
```
Comment (akx) [inline shoop/xtheme/plugins/widgets.py]: If there are 15,000 products, this will iterate over all of them, then pick one. (I believe we have a helper called `first()` somewhere...)

Also, there's nothing guaranteeing pks are integers; comparing as strings would be safer.

It would really be better, all in all, that the ChoiceField passed the selected object itself down to the widget for rendering.

File: shoop/xtheme/plugins/widgets.py [hunk 2]
```diff
@@ -90,3 +91,38 @@ def decompress(self, value):
...
+            **kwargs
+        )
+
+    def _get_custom_choices(self):
+        return XThemeModelChoiceIterator(self)
+
+    choices = property(_get_custom_choices, forms.ChoiceField._set_choices)
+
+
+class XThemeModelChoiceIterator(forms.models.ModelChoiceIterator):
```
Comment (akx) [inline shoop/xtheme/plugins/widgets.py]: Why is this necessary?

File: shoop/xtheme/plugins/widgets.py [hunk 3]
```diff
@@ -90,3 +91,38 @@ def decompress(self, value):
...
+            render_to_string("shoop/xtheme/_model_widget.jinja", {
+                "name": name,
+                "selected_object": [object for pk, object in self.choices if pk == int(value)][0] if value else None,
+                "objects": self.choices,
+            })
+        )
+
+
+class XThemeModelChoiceField(forms.ModelChoiceField):
+    def __init__(self, queryset, empty_label="---------", required=True, label=None, initial=None, *args, **kwargs):
```
Comment (akx) [inline shoop/xtheme/plugins/widgets.py]: Would probably be better to just `*args, **kwargs` here (or really just `**kwargs` should be enough given the use).

**Comment by Pikkupomo** [comment]
wrong ref @tulimaki?

File: shoop/xtheme/plugins/widgets.py (hunk 1 above)
Comment (akx) [inline shoop/xtheme/plugins/widgets.py]: ... Oh, wait, you're only using this to render `selected`?

Why not just pass `value` down to the template context and do `if value == object.pk` in the render loop there?

File: shoop/xtheme/templates/shoop/xtheme/_model_widget.jinja [hunk 4]
```diff
@@ -0,0 +1,20 @@
+<select id="id_{{ name }}" name="{{ name }}" class="xtheme-model-choice-widget">
+    {% for pk, object in objects %}
+        {% if pk %}
+            {% set url = shoop_admin.model_url(object, "detail") %}
```
Comment (akx) [inline shoop/xtheme/templates/shoop/xtheme/_model_widget.jinja]: Jinja supports ternary expressions; `{% set url = (shoop_admin.model_url(...) if pk else "") %}` is arguably clearer.

File: shoop/xtheme/plugins/widgets.py (hunk 2 above)
Comment (tulimaki) [inline shoop/xtheme/plugins/widgets.py]: It's not I think that overriding label_from_instance could be better solution for this.

File: shoop/xtheme/plugins/widgets.py (hunk 1 above)
Comment (tulimaki) [inline shoop/xtheme/plugins/widgets.py]: correct

File: shoop/xtheme/plugins/widgets.py [hunk 5]
```diff
@@ -90,3 +91,27 @@ def decompress(self, value):
...
+                "objects": self.choices,
+            })
+        )
+
+
+class XThemeModelChoiceField(forms.ModelChoiceField):
+    widget = XThemeModelChoiceWidget
+
+    def __init__(self, **kwargs):
+        super(XThemeModelChoiceField, self).__init__(**kwargs)
```
Comment (akx) [inline shoop/xtheme/plugins/widgets.py]: Unnecessarily overridden init?

Comment (tulimaki) [inline shoop/xtheme/plugins/widgets.py]: yes

File: shoop/xtheme/plugins/widgets.py [hunk 6]
```diff
@@ -90,3 +91,24 @@ def decompress(self, value):
...
+            })
+        )
+
+
+class XThemeModelChoiceField(forms.ModelChoiceField):
+    widget = XThemeModelChoiceWidget
+
+    def label_from_instance(self, obj):
+        """
+        Use objects in choices instead of strings
```
Comment (akx) [inline shoop/xtheme/plugins/widgets.py]: This is a comment, not a docstring.

**Comment by tulimaki** [comment]
retest this please
